File: login/load.js
const CDP = require('chrome-remote-interface');
const fs = require('fs')
const parse = require('url').parse
const path = require('path')

let cookieReq = [];
let compare = { 'same': {}, 'diff': {} };
let reqheaders = {};
let filename = '';
let firstid = 0;
let turned = false;
let resource = {};
let currentUrl = '';

function ReadReq(filename) {
    let data = fs.readFileSync(path.join('headers', 'login', filename.replace(/\//g, '_') + '.json'), 'utf-8');
    const reqJson = JSON.parse(data).available;
    for (let key in reqJson) {
        let req = reqJson[key];
        if (req.method == 'GET' && 'headers' in req && ('cookie' in req.headers || 'Cookie' in req.headers) && 'bytes' in req)
            cookieReq.push(req);
    }
}

function ReadResource(filename) {
    let data = fs.readFileSync(path.join('headers', 'resources', filename.replace(/\//g, '_') + '.json'), 'utf-8');
    resource = JSON.parse(data);
}

function identical(req, body, resource){
    if (['XHR', 'Document', 'Fetch'].indexOf(req['type']) >= 0 )
        return req.bytes == body.length;
    return body == resource[req.url].body;
}

CDP((client) => {
    // extract domains
    const { Network, Page, Security } = client;
    // console.log(Security);

    Security.setIgnoreCertificateErrors({ ignore: true });
    //Security.disable();

    Network.requestWillBeSent(params => {
        //console.log("request: " + params.requestId);
        
        if (params.documentURL == currentUrl && !turned){
            firstid = params.requestId;
            turned = true;
        }
     });

    Network.responseReceived(params => {
        // console.log("response: " + params.requestId);
        if (params.requestId == firstid) {
            let decision = 'headers' in params.response;
            // Check download behavior
            decision = decision && ('content-disposition' in params.response.headers 
                            || ('content-type' in params.response.headers && params.response.headers['content-type'].indexOf('octet-stream') != -1));
            // Checkfor content-length (0 means no load event)
            decision = decision || ('Content-Length' in params.response.headers && params.response.headers['Content-Length'] == 0) 
                                || ('content-length' in params.response.headers && params.response.headers['content-length'] == 0)
            if (decision) {
                // console.log("AHA");
                cookieReq.shift();
                if (!cookieReq.length) {
                    client.close();
                    fs.writeFile(path.join('headers', 'compare', filename.replace(/\//g, '_') + '.json'), JSON.stringify(compare), 'utf-8', err => {
                        if (err) throw err;
                    });
                } else {
                    currentUrl = cookieReq[0].url;
                    // console.log("Finished: "+ currentUrl + '\n\n');
                    firstid = 0;
                    turned = false;
                    return Page.navigate({ url: currentUrl });
                }
                return;
            }
            reqheaders[params.requestId] = params.response.requestHeaders;
            // firstid = params.requestId;
            // turned = true;
        }
    });

    Network.loadingFinished(async (params) => {
        //console.log("load: " + params.requestId);
        if (params.requestId == firstid) {
            Page.stopLoading();
            let bytes = 0;
            let data = ''
            try {
                data = await Network.getResponseBody({ requestId: firstid });
                data = data.body;
                bytes = data.length;
            } catch (err) {
            }
            // console.log(bytes);
            if (identical(cookieReq[0], data, resource))
                compare['same'][cookieReq[0].url] = [
                    bytes,
                    cookieReq[0].type,
                    cookieReq[0].initiator
                ];
            else
                compare['diff'][cookieReq[0].url] = {
                    bytes: [bytes, cookieReq[0].bytes, cookieReq[0].type, cookieReq[0].initiator],
                    headers: reqheaders[params.requestId]
                };
            cookieReq.shift();
            if (!cookieReq.length) {
                client.close();
                fs.writeFile(path.join('headers', 'compare', filename.replace(/\//g, '_') + '.json'), JSON.stringify(compare), 'utf-8', err => {
                    if (err) throw err;
                });
            } else {
                currentUrl = cookieReq[0].url;
                // console.log("\n\nFinished: " + currentUrl + '\n\n');
                firstid = 0;
                turned = false;
                return Page.navigate({ url: currentUrl });
            }
        }
    });

    // Page.loadEventFired(() => {
    //     cookieReq.shift();
    //     if (!cookieReq.length) {
    //         client.close();
    //         fs.writeFile(path.join('headers', 'compare', filename + '.json'), JSON.stringify(compare), 'utf-8', err => {
    //             if (err) throw err;
    //         });
    //     } else {
    //         url = cookieReq[0].url;
    //         firstid = 0;
    //         turned = false;
    //         console.log(url);
    //         return Page.navigate({ url: url });
    //     }
    // });

    // enable events then start!
    Promise.all([
        Network.enable(),
        Network.clearBrowserCache(),
        Page.enable(),
        Page.setDownloadBehavior({ behavior: 'deny' })
    ]).then(() =>{
        filename = process.argv[2];
        ReadReq(filename);
        ReadResource(filename);
        currentUrl = cookieReq[0].url;
        return Page.navigate({ url: currentUrl });
    }).catch((err) => {
        console.error(err);
        client.close();
    });

}).on('error', (err) => {
    // cannot connect to the remote endpoint
    console.error(err);
});

File: login/load.py
#!/anaconda3/bin/python3
"""
load the resouces with unloggedin (empty) cookie
"""
import os
import sys
import json
from subprocess import *
import requests
import common
from urllib.parse import urlparse
from hyper.contrib import HTTP20Adapter
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics.pairwise import cosine_similarity

host = sys.argv[1]
url = 'https://' + host


def load(login):
    compare = {'same': {}, 'diff': {}}
    for _, req in login.items():
        if "bytes" not in req:
            continue
        if req['method'] == 'GET' and 'headers' in req and ('cookie' in req['headers'] or 'Cookie' in req['headers']):
            try:
                if 'cookie' in req['headers']:
                    del req['headers']['cookie']
                else:
                    del req['headers']['Cookie']
                # s = requests.Session()
                # s.mount("https://" + urlparse(req['url']).netloc, HTTP20Adapter())
                # r = s.get(req['url'], headers=req['headers'], timeout=5)
                headers = common.strip_colon(req['headers'])
                r = requests.get(req['url'], headers=headers, timeout=10)
            except Exception as e: 
                print(str(e))
                continue
            byte, text = common.find_length(r, True)
            is_similar, similarity = common.similar(byte, req['bytes'], text, check_output(['python3', 'extract_resource.py', host.replace('/', '_'), req['url']]).decode())
            if is_similar:
                compare['same'][req['url']] = [byte]
            else:
                compare['diff'][req['url']] = [byte, req['bytes'], similarity]
    f = open(os.path.join('headers', 'compare', host.replace('/', '_') + '.json'), 'w+')
    f.write(json.dumps(compare))

def identical_response(byte, body, req, resource):
    if req['type'] in ['Document', 'XHR', 'Fetch', 'Script']:
        return byte == req['bytes']
    return body == resource['body']



# def find_length(r):
#     if 'content-encoding' in r.headers and r.headers['content-encoding'] == 'br':
#         return (len(brotli.decompress(r.content)),  brotli.decompress(r.content))
#     else:
#         return (len(base64.b64encode(r.content)), base64.b64encode(r.content).decode())


f = open(os.path.join('headers', 'login', host.replace('/', '_') + '.json'), 'r').read()
login_json = json.loads(f)
load(login_json['available'])
